test/perf/perf.test.js

```diff
@@ -0,0 +1,116 @@
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+/* eslint-disable no-undef */
+/* eslint-disable camelcase */
+const { fetch } = require('@adobe/helix-fetch');
+const assert = require('assert');
+const { getUrls, parseTiming } = require('../utils');
+
+const baseService = process.env.HLX_FASTLY_NAMESPACE;
+const testService = process.env.TEST_SERVICE;
+/**
+ * generates strings with information about changes in rendering time
+ *
+ * @param {*} timeObj1 object of pipeline step times and desc
+ * @param {*} timeObj2 object of pipeline step times and desc
+ */
+function testTiming(timeObj1, timeObj2) {
+  const actual = Object.keys(timeObj2).reduce((prev, curr) => {
+    if (curr in timeObj1) {
+      const baseTime = timeObj1[curr];
+      const branchTime = timeObj2[curr];
+      // eslint-disable-next-line no-param-reassign
+      prev += `commit will ${baseTime < branchTime ? 'INCREASE' : 'DECREASE'} time of step: ${curr} by ${Number(branchTime - baseTime).toFixed(4)}\n`;
+    } else {
+      // eslint-disable-next-line no-param-reassign
+      prev += `commit adds step ${curr}: ${branchTime}ms\n`;
+    }
+    return prev;
+  }, '\n');
+
+  const expected = Object.keys(timeObj1).reduce((prev, curr) => {
+    if (curr in timeObj2) {
+      const baseTime = timeObj1[curr];
+      // eslint-disable-next-line no-param-reassign
+      prev += `commit removes step ${curr}: ${baseTime}ms\n`;
+    }
+    return prev;
+  }, actual);
+
+  return { actual, expected };
+}
+
+/**
+ * constructs list of promises of calls to fetch
+ *
+ */
+async function sendPerfRequests() {
+  const urls = await getUrls();
+  const baseHeaders = {
+    'X-DEBUG': baseService,
+    'cache-control': 'no-cache',
+  };
+  const testHeaders = {
+    'X-DEBUG': testService,
+    'cache-control': 'no-cache',
+  };
+  const visits = urls.reduce((prev, curr) => {
+    prev.push(fetch(`${curr.base}`, { headers: baseHeaders }).then((res) => {
+      // consume response
+      res.text();
+      return res;
+    }));
+    prev.push(fetch(`${curr.branch}`, { headers: testHeaders }).then((res) => {
+      // consume response
+      res.text();
+      return res;
+    }));
+    return prev;
+  }, []);
+
+  return Promise.all(visits);
+}
+/**
+ * runs performance tests and compares metrics
+ *
+ */
+async function runPerfTests(res) {
+  describe('pipeline rendering time', () => {
+    for (let idx = 0; idx < res.length; idx += 2) {
+      const { url } = res[idx];
+      const baseTime = res[idx].headers.get('server-timing');
+      const branchTime = res[idx + 1].headers.get('server-timing');
+      if (baseTime && branchTime) {
+        const baseTimeObj = parseTiming(baseTime);
+        const branchTimeObj = parseTiming(branchTime);
+        it(`testing pipeline rendering time for ${url}`, () => {
+          try {
+            assert.deepEqual(baseTimeObj, branchTimeObj);
```

Comment:
I don't think that the timings are ever exact equal.....

Reply:
Right; what do you suggest? Maybe we can say the difference can be within a certain range as long as
 |testExecTime - baseExecTime| <= threshold and if it's greater than we sound the alarms? or I can just change this so that it just is extra information for hlx devs but not tests that must pass or be ignored?

Reply:
yeah..lets add some diff as information for now....